### PR TITLE
HTTP Validator API: `/eth/v1/validator/{pubkey}/gas_limit`

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**

Other

**What does this PR do? Why is it needed?**

This PR implements the pure http REST representation of `/eth/v1/validator/{pubkey}/gas_limit` with `GET`,`POST` and `DELETE`

execution examples
```
james@Jamess-MacBook-Pro-2 prysm % curl -X GET "http://127.0.0.1:7500/eth/v1/validator/0x93247f2209abcacf57b75a51dafae777f9dd38bc7053d1af526f220a7489a6d3a2753e5f3e8b1cfe39b56f43611df74a/gas_limit" -H "accept: application/json"
{"data":{"pubkey":"0x93247f2209abcacf57b75a51dafae777f9dd38bc7053d1af526f220a7489a6d3a2753e5f3e8b1cfe39b56f43611df74a","gas_limit":"30000000"}}
james@Jamess-MacBook-Pro-2 prysm % curl -X POST http://127.0.0.1:7500/eth/v1/validator/0x93247f2209abcacf57b75a51dafae777f9dd38bc7053d1af526f220a7489a6d3a2753e5f3e8b1cfe39b56f43611df74a/gas_limit \
     -H "Content-Type: application/json" \
     -d '{"gas_limit":"4000"}'
james@Jamess-MacBook-Pro-2 prysm % curl -X GET "http://127.0.0.1:7500/eth/v1/validator/0x93247f2209abcacf57b75a51dafae777f9dd38bc7053d1af526f220a7489a6d3a2753e5f3e8b1cfe39b56f43611df74a/gas_limit" -H "accept: application/json"                                                 
{"data":{"pubkey":"0x93247f2209abcacf57b75a51dafae777f9dd38bc7053d1af526f220a7489a6d3a2753e5f3e8b1cfe39b56f43611df74a","gas_limit":"4000"}}
james@Jamess-MacBook-Pro-2 prysm % curl -X DELETE "http://127.0.0.1:7500/eth/v1/validator/0x93247f2209abcacf57b75a51dafae777f9dd38bc7053d1af526f220a7489a6d3a2753e5f3e8b1cfe39b56f43611df74a/gas_limit" -H  "accept: application/json" -H  "Content-Type: application/json" 
james@Jamess-MacBook-Pro-2 prysm % curl -X GET "http://127.0.0.1:7500/eth/v1/validator/0x93247f2209abcacf57b75a51dafae777f9dd38bc7053d1af526f220a7489a6d3a2753e5f3e8b1cfe39b56f43611df74a/gas_limit" -H "accept: application/json"                                          
{"data":{"pubkey":"0x93247f2209abcacf57b75a51dafae777f9dd38bc7053d1af526f220a7489a6d3a2753e5f3e8b1cfe39b56f43611df74a","gas_limit":"30000000"}}


```
**Which issues(s) does this PR fix?**

Part of #13041

### CODE REVIEW

File: validator/rpc/handlers_keymanager.go [hunk 1]
```diff
@@ -238,3 +241,169 @@ func (s *Server) DeleteRemoteKeys(w http.ResponseWriter, r *http.Request) {
...
+	pubkey, valid := shared.ValidateHex(w, "pubkey", rawPubkey, fieldparams.BLSPubkeyLength)
+	if !valid {
+		return
+	}
+
+	resp := &GetGasLimitResponse{
+		Data: &GasLimitMetaData{
+			Pubkey: rawPubkey,
+		},
+	}
```
Comment (saolyn) [inline validator/rpc/handlers_keymanager.go:270]: Could we set the default gas limit first:
```suggestion
	resp := &GetGasLimitResponse{
		Data: &GasLimitMetaData{
			Pubkey: rawPubkey,
			GasLimit: fmt.Sprintf("%d", params.BeaconConfig().DefaultBuilderGasLimit),
		},
	}
```
Then override it if settings are not nil, clear out all the writes and returns and just have them at the very end?
If my explanation is a bit confusing I mean do this:

```
resp := &GetGasLimitResponse{
    Data: &GasLimitMetaData{
	Pubkey:   rawPubkey,
	GasLimit: fmt.Sprintf("%d", params.BeaconConfig().DefaultBuilderGasLimit),
    },
}
settings := s.validatorService.ProposerSettings()
if settings != nil {
	proposerOption, found := settings.ProposeConfig[bytesutil.ToBytes48(pubkey)]
	if found {
		if proposerOption.BuilderConfig != nil {
			resp.Data.GasLimit = fmt.Sprintf("%d", proposerOption.BuilderConfig.GasLimit)
		}
	} else if settings.DefaultConfig != nil && settings.DefaultConfig.BuilderConfig != nil {
		resp.Data.GasLimit = fmt.Sprintf("%d", s.validatorService.ProposerSettings().DefaultConfig.BuilderConfig.GasLimit)
	}
}
http2.WriteJson(w, resp)
```

File: validator/rpc/handlers_keymanager.go [hunk 2]
```diff
@@ -238,3 +241,169 @@ func (s *Server) DeleteRemoteKeys(w http.ResponseWriter, r *http.Request) {
...
+		http2.HandleError(w, "Could not set proposer settings: "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.WriteHeader(http.StatusAccepted)
+}
+
+// DeleteGasLimit deletes the gas limit by public key
+func (s *Server) DeleteGasLimit(w http.ResponseWriter, r *http.Request) {
+	ctx, span := trace.StartSpan(r.Context(), "validator.keymanagerAPI.SetGasLimit")
```
Comment (saolyn) [inline validator/rpc/handlers_keymanager.go:365]:
```suggestion
	ctx, span := trace.StartSpan(r.Context(), "validator.keymanagerAPI.DeleteGasLimit")
```

File: validator/rpc/handlers_keymanager.go [hunk 3]
```diff
@@ -238,3 +241,169 @@ func (s *Server) DeleteRemoteKeys(w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	gasLimit, valid := shared.ValidateUint(w, "Gas Limit", req.GasLimit)
+	if !valid {
+		return
+	}
+
+	settings := s.validatorService.ProposerSettings()
+	if settings == nil {
+		http2.HandleError(w, "no proposer settings were found to update", http.StatusInternalServerError)
```
Comment (saolyn) [inline validator/rpc/handlers_keymanager.go:324]: These should all be capitalized
```suggestion
		http2.HandleError(w, "No proposer settings were found to update", http.StatusInternalServerError)
```

File: validator/rpc/handlers_keymanager.go [hunk 4]
```diff
@@ -238,3 +241,169 @@ func (s *Server) DeleteRemoteKeys(w http.ResponseWriter, r *http.Request) {
...
+		return
+	} else if settings.ProposeConfig == nil {
+		if settings.DefaultConfig == nil || settings.DefaultConfig.BuilderConfig == nil || !settings.DefaultConfig.BuilderConfig.Enabled {
+			http2.HandleError(w, "gas limit changes only apply when builder is enabled", http.StatusInternalServerError)
```
Comment (saolyn) [inline validator/rpc/handlers_keymanager.go:328]:
```suggestion
			http2.HandleError(w, "Gas limit changes only apply when builder is enabled", http.StatusInternalServerError)
```

File: validator/rpc/handlers_keymanager.go [hunk 5]
```diff
@@ -238,3 +241,169 @@ func (s *Server) DeleteRemoteKeys(w http.ResponseWriter, r *http.Request) {
...
+		}
+		settings.ProposeConfig = make(map[[fieldparams.BLSPubkeyLength]byte]*validatorServiceConfig.ProposerOption)
+		option := settings.DefaultConfig.Clone()
+		option.BuilderConfig.GasLimit = validator.Uint64(gasLimit)
+		settings.ProposeConfig[bytesutil.ToBytes48(pubkey)] = option
+	} else {
+		proposerOption, found := settings.ProposeConfig[bytesutil.ToBytes48(pubkey)]
+		if found {
+			if proposerOption.BuilderConfig == nil || !proposerOption.BuilderConfig.Enabled {
+				http2.HandleError(w, "gas limit changes only apply when builder is enabled", http.StatusInternalServerError)
```
Comment (saolyn) [inline validator/rpc/handlers_keymanager.go:339]:
```suggestion
				http2.HandleError(w, "Gas limit changes only apply when builder is enabled", http.StatusInternalServerError)
```

File: validator/rpc/handlers_keymanager.go [hunk 6]
```diff
@@ -238,3 +241,169 @@ func (s *Server) DeleteRemoteKeys(w http.ResponseWriter, r *http.Request) {
...
+				return
+			} else {
+				proposerOption.BuilderConfig.GasLimit = validator.Uint64(gasLimit)
+			}
+		} else {
+			if settings.DefaultConfig == nil {
+				http2.HandleError(w, "gas limit changes only apply when builder is enabled", http.StatusInternalServerError)
```
Comment (saolyn) [inline validator/rpc/handlers_keymanager.go:346]:
```suggestion
				http2.HandleError(w, "Gas limit changes only apply when builder is enabled", http.StatusInternalServerError)
```

File: validator/rpc/handlers_keymanager.go [hunk 7]
```diff
@@ -238,3 +241,169 @@ func (s *Server) DeleteRemoteKeys(w http.ResponseWriter, r *http.Request) {
...
+			}
+			// Successfully deleted gas limit (reset to proposer config default or global default).
+			// Return with success http code "204".
+			w.WriteHeader(http.StatusNoContent)
+			return
+		}
+	}
+	// Otherwise, either no proposerOption is found for the pubkey or proposerOption.BuilderConfig is not enabled at all,
+	// we respond "not found".
+	http2.HandleError(w, fmt.Sprintf("no gaslimt found for pubkey: %q", rawPubkey), http.StatusNotFound)
```
Comment (saolyn) [inline validator/rpc/handlers_keymanager.go:408]:
```suggestion
	http2.HandleError(w, fmt.Sprintf("No gaslimt found for pubkey: %q", rawPubkey), http.StatusNotFound)
```

File: validator/rpc/server.go [hunk 8]
```diff
@@ -214,6 +213,26 @@ func (s *Server) Start() {
...
+	s.router.HandleFunc("/eth/v1/remotekeys", s.ListRemoteKeys).Methods(http.MethodGet)
+	s.router.HandleFunc("/eth/v1/remotekeys", s.ImportRemoteKeys).Methods(http.MethodPost)
+	s.router.HandleFunc("/eth/v1/remotekeys", s.DeleteRemoteKeys).Methods(http.MethodDelete)
+	s.router.HandleFunc("/eth/v1/validator/{pubkey}/gas_limit", s.GetGasLimit).Methods(http.MethodGet)
+	s.router.HandleFunc("/eth/v1/validator/{pubkey}/gas_limit", s.SetGasLimit).Methods(http.MethodPost)
+	s.router.HandleFunc("/eth/v1/validator/{pubkey}/gas_limit", s.DeleteGasLimit).Methods(http.MethodDelete)
+	s.router.HandleFunc("/eth/v1/validator/{pubkey}/voluntary_exit", s.SetVoluntaryExit).Methods(http.MethodPost)
+	// ...
+
+	return nil // or any error that might occur during initialization
```
Comment (saolyn) [inline validator/rpc/server.go:233]: Wasn't this already added in your fix a couple days ago?

File: validator/rpc/handlers_keymanager.go [hunk 9]
```diff
@@ -238,3 +241,169 @@ func (s *Server) DeleteRemoteKeys(w http.ResponseWriter, r *http.Request) {
...
+			resp.Data.GasLimit = fmt.Sprintf("%d", s.validatorService.ProposerSettings().DefaultConfig.BuilderConfig.GasLimit)
+			http2.WriteJson(w, resp)
+			return
+		}
+	}
+	resp.Data.GasLimit = fmt.Sprintf("%d", params.BeaconConfig().DefaultBuilderGasLimit)
+	http2.WriteJson(w, resp)
+}
+
+// SetGasLimit updates GasLimt of the public key.
```
Comment (saolyn) [inline validator/rpc/handlers_keymanager.go:290]:
```suggestion
// SetGasLimit updates the GasLimit of the provided public key.
```

File: validator/rpc/handlers_keymanager.go (hunk 1 above)
Comment (james-prysm) [reply validator/rpc/handlers_keymanager.go:270]: yup we can do that

File: validator/rpc/handlers_keymanager.go [hunk 10]
```diff
@@ -238,3 +241,169 @@ func (s *Server) DeleteRemoteKeys(w http.ResponseWriter, r *http.Request) {
...
+		},
+	}
+	settings := s.validatorService.ProposerSettings()
+	if settings != nil {
+		proposerOption, found := settings.ProposeConfig[bytesutil.ToBytes48(pubkey)]
+		if found {
+			if proposerOption.BuilderConfig != nil {
+				resp.Data.GasLimit = fmt.Sprintf("%d", proposerOption.BuilderConfig.GasLimit)
+				http2.WriteJson(w, resp)
+				return
```
Comment (rkapka) [inline validator/rpc/handlers_keymanager.go:279]: you don't need this, execution will flow to the last line

File: validator/rpc/handlers_keymanager.go [hunk 11]
```diff
@@ -238,3 +241,169 @@ func (s *Server) DeleteRemoteKeys(w http.ResponseWriter, r *http.Request) {
...
+			}
+		} else if settings.DefaultConfig != nil && settings.DefaultConfig.BuilderConfig != nil {
+			resp.Data.GasLimit = fmt.Sprintf("%d", s.validatorService.ProposerSettings().DefaultConfig.BuilderConfig.GasLimit)
+			http2.WriteJson(w, resp)
+			return
```
Comment (rkapka) [inline validator/rpc/handlers_keymanager.go:284]: same here

File: validator/rpc/handlers_keymanager.go [hunk 12]
```diff
@@ -238,3 +241,169 @@ func (s *Server) DeleteRemoteKeys(w http.ResponseWriter, r *http.Request) {
...
+			}
+			// Successfully deleted gas limit (reset to proposer config default or global default).
+			// Return with success http code "204".
+			w.WriteHeader(http.StatusNoContent)
+			return
+		}
+	}
+	// Otherwise, either no proposerOption is found for the pubkey or proposerOption.BuilderConfig is not enabled at all,
+	// we respond "not found".
+	http2.HandleError(w, fmt.Sprintf("No gaslimt found for pubkey: %q", rawPubkey), http.StatusNotFound)
```
Comment (rkapka) [inline validator/rpc/handlers_keymanager.go:408]:
```suggestion
	http2.HandleError(w, fmt.Sprintf("No gas limit found for pubkey: %q", rawPubkey), http.StatusNotFound)
```

File: validator/rpc/server.go [hunk 13]
```diff
@@ -214,6 +213,26 @@ func (s *Server) Start() {
...
+	s.router.HandleFunc("/eth/v1/remotekeys", s.ListRemoteKeys).Methods(http.MethodGet)
+	s.router.HandleFunc("/eth/v1/remotekeys", s.ImportRemoteKeys).Methods(http.MethodPost)
+	s.router.HandleFunc("/eth/v1/remotekeys", s.DeleteRemoteKeys).Methods(http.MethodDelete)
+	s.router.HandleFunc("/eth/v1/validator/{pubkey}/gas_limit", s.GetGasLimit).Methods(http.MethodGet)
+	s.router.HandleFunc("/eth/v1/validator/{pubkey}/gas_limit", s.SetGasLimit).Methods(http.MethodPost)
+	s.router.HandleFunc("/eth/v1/validator/{pubkey}/gas_limit", s.DeleteGasLimit).Methods(http.MethodDelete)
+	s.router.HandleFunc("/eth/v1/validator/{pubkey}/voluntary_exit", s.SetVoluntaryExit).Methods(http.MethodPost)
+	// ...
+	log.Info("Initialized REST API routes")
+	return nil // or any error that might occur during initialization
```
Comment (rkapka) [inline validator/rpc/server.go:233]:
```suggestion
	return nil
```

**Review by saolyn — APPROVED** [review]
LGTM

**Review by rkapka — DISMISSED** [review]

**Pull request merged**